### PR TITLE
KTOR-2931 Support streaming of body in curl

### PR DESCRIPTION
**Pull request by rsinukov**

### CODE REVIEW

File: ktor-io/js/src/io/ktor/utils/io/ByteReadChannelJs.kt [hunk 1]
```diff
@@ -167,6 +167,8 @@ public actual interface ByteReadChannel {
      */
     public actual suspend fun discard(max: Long): Long
 
+    public actual suspend fun awaitContent()
```
Comment (e5l) [inline ktor-io/js/src/io/ktor/utils/io/ByteReadChannelJs.kt:170]: please add KDocs

File: ktor-io/jvm/src/io/ktor/utils/io/ByteChannelSequentialJVM.kt [hunk 2]
```diff
@@ -86,10 +86,6 @@ public class ByteChannelSequentialJVM(
     }
 
     override fun readAvailable(min: Int, block: (ByteBuffer) -> Unit): Int {
-        if (closed) {
-            throw closedCause ?: ClosedSendChannelException("Channel closed for read")
```
Comment (e5l) [inline ktor-io/jvm/src/io/ktor/utils/io/ByteChannelSequentialJVM.kt:90]: It looks like we should throw `closedCause` if it's not null

File: ktor-io/posix/src/io/ktor/utils/io/ByteReadChannelNative.kt [hunk 3]
```diff
@@ -43,6 +43,22 @@ public actual interface ByteReadChannel {
...
+     * Warning: it is not guaranteed that all of available bytes will be represented as a single byte buffer
+     * eg: it could be 4 bytes available for read but the provided byte buffer could have only 2 available bytes:
+     * in this case you have to invoke read again (with decreased [min] accordingly).
+     *
+     * @param min amount of bytes available for read, should be positive
+     * @param block to be invoked when at least [min] bytes available
+     *
+     * @return number of consumed bytes or -1 if the block wasn't executed.
+     */
+    public fun readAvailable(min: Int, block: (Buffer) -> Unit): Int
```
Comment (e5l) [inline ktor-io/posix/src/io/ktor/utils/io/ByteReadChannelNative.kt:60]: KDocs should be added to the actual implementations as well

**Review by e5l — CHANGES_REQUESTED** [review]

File: ktor-io/jvm/src/io/ktor/utils/io/ByteChannelSequentialJVM.kt (hunk 2 above)
Comment (rsinukov) [reply ktor-io/jvm/src/io/ktor/utils/io/ByteChannelSequentialJVM.kt:90]: thanks, don't know why I lost it. fixed!

File: ktor-client/ktor-client-curl/posix/src/io/ktor/client/engine/curl/CurlProcessor.kt [hunk 4]
```diff
@@ -20,72 +18,45 @@ private lateinit var curlApi: CurlMultiApiHandler
 internal class CurlProcessor(
     override val coroutineContext: CoroutineContext
 ) : CoroutineScope {
-    private val worker: Worker = Worker.start()
-    private val responseConsumers: ConcurrentMap<CurlRequestData, CompletableDeferred<CurlSuccess>> = ConcurrentMap()
-    private val activeRequests = atomic(0)
+    @OptIn(ExperimentalCoroutinesApi::class)
+    private val curlDispatcher: CoroutineDispatcher = newSingleThreadContext("curl-dispatcher")
```
Comment (e5l) [inline ktor-client/ktor-client-curl/posix/src/io/ktor/client/engine/curl/CurlProcessor.kt:22]: It looks like we should explicitly close the dispatcher

File: ktor-client/ktor-client-curl/posix/src/io/ktor/client/engine/curl/CurlProcessor.kt [hunk 5]
```diff
@@ -20,72 +18,45 @@ private lateinit var curlApi: CurlMultiApiHandler
...
         } finally {
             requestCleaner.dispose()
         }
     }
 
     public fun close() {
-        worker.execute(TransferMode.SAFE, { Unit }) { curlApi.close() }
-        worker.requestTermination()
+        curlScope.launch { curlApi.close() }
+        curlScope.cancel()
```
Comment (e5l) [inline ktor-client/ktor-client-curl/posix/src/io/ktor/client/engine/curl/CurlProcessor.kt:52]: it looks like `launch` and `cancel` are in the race.

**Review by e5l — CHANGES_REQUESTED** [review]

File: ktor-client/ktor-client-curl/posix/src/io/ktor/client/engine/curl/CurlProcessor.kt (hunk 5 above)
Comment (rsinukov) [reply ktor-client/ktor-client-curl/posix/src/io/ktor/client/engine/curl/CurlProcessor.kt:55]: thanks, updated

**Review by e5l — APPROVED** [review]
lgtm

**Comment by e5l** [comment]
Good job, Please check linux native tests before merging

**Pull request merged**